File: android/app/src/main/kotlin/com/example/hotel_reception/procesor/PersonRandomizer.kt
package com.example.hotel_reception.procesor

import com.example.hotel_reception.procesor.model.Person
import com.example.hotel_reception.procesor.model.RoomType
import com.example.hotel_reception.procesor.model.Season
import java.util.*
import kotlin.math.roundToInt
import kotlin.random.Random

class PersonRandomizer {
    companion object {
        private fun randomChanceToBreak() = Random.nextDouble(0.0, 3.0)
        private fun randomChanceToCompletelyBreak() = Random.nextDouble(0.0, 1.0)
        private fun randomDaysAmount() = Random.nextInt(1, 5)
        private fun randomPersonAmount() = Random.nextInt(0, 10)

        fun generatePersonQuery(date: Date): List<Person> {
            val personCount = randomPersonAmount()
            val persons = mutableListOf<Person>()
            for (i in 0..personCount) {
                val days = randomDaysAmount()
                persons.add(
                    Person(
                        days,
                        date,
                        Calendar.getInstance().apply {
                            time = date
                            add(Calendar.DATE, days)
                        }.time,
                        randomChanceToBreak().roundToInt(),
                        randomChanceToCompletelyBreak().roundToInt(),
                        roomPreferences = generatePersonPreferences()
                    )
                )
            }
            return persons
        }

        private fun generatePersonPreferences(): List<RoomType> {
            val chance = Random.nextInt(0, 100)
            return when (chance) {
                in 71..85 -> listOf(RoomType.econom, RoomType.suite)
                in 76..100 -> listOf(RoomType.econom, RoomType.suite, RoomType.president)
                else -> listOf(RoomType.econom)
            }
        }
    }
}

File: ios/Runner/Processor/event/PersonRandomizer.swift
//
//  PersonRandomizer.swift
//  Runner
//
//  Created by Serhii Rakovskyi on 18.08.2022.
//

import Foundation

class PersonRandomizer {
    private func randomChanceToBreak() -> Double {
        return Double.random(in: 0.0...3.3)
    }
    
    private func randomChanceToCompletelyBreak() -> Double {
        return Double.random(in: 0.0...1.0)
    }
    private func randomDaysAmount() -> Int {
        return Int.random(in: 1...5)
    }
    private func randomPersonAmount() -> Int{
        return Int.random(in: 0...10)
    }
    
    func generatePersonQuery(date: Date) -> Array<Person> {
        let personCount = randomPersonAmount()
        var persons:Array<Person> = []
        for _ in 0...personCount {
            let days = randomDaysAmount()
            var dayComponent = DateComponents()
            dayComponent.day = 1
            persons.append(Person(
                daysAmount: days, startDate: date, endDate: Calendar.current.date(byAdding: dayComponent, to: date)!, chanceToReduceRoomLifeTime: Int(randomChanceToBreak()), chanceToBreakRoom: Int(randomChanceToCompletelyBreak()), roomPreferences: generatePersonPreferences()
            ))
        }
        return persons
    }
    
    private func generatePersonPreferences() ->  Array<RoomType> {
        let chance = Int.random(in: 0...100)
        switch chance {
        case  71...85: return [RoomType.econom(40, 1500), RoomType.suite(95, 2500)]
        case  76...100: return [RoomType.econom(40, 1500), RoomType.suite(95, 2500), RoomType.president(150, 3500)]
        default: return [RoomType.econom(40, 1500)]
            
        }
    }
}
